GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m

```diff
@@ -140,110 +147,132 @@ - (void)uploadTarget:(GDTCORTarget)target withConditions:(GDTCORUploadConditions
 
   // 1. Check if the conditions for the target are suitable.
   [self isReadyToUploadTarget:target conditions:conditions]
+      .validateOn(self.uploaderQueue,
+                  ^BOOL(id result) {
+                    // 2. Stop the operation if it has been cancelled.
+                    return !self.isCancelled;
+                  })
       .thenOn(self.uploaderQueue,
               ^id(id result) {
-                // 2. Remove previously attempted batches
+                // 3. Remove previously attempted batches.
                 return [storage removeAllBatchesForTarget:target deleteEvents:NO];
               })
       .thenOn(self.uploaderQueue,
               ^FBLPromise<NSNumber *> *(id result) {
                 // There may be a big amount of events stored, so creating a batch may be an
                 // expensive operation.
 
-                // 3. Do a lightweight check if there are any events for the target first to
+                // 4. Do a lightweight check if there are any events for the target first to
                 // finish early if there are no.
                 return [storage hasEventsForTarget:target];
               })
       .validateOn(self.uploaderQueue,
                   ^BOOL(NSNumber *hasEvents) {
-                    // Stop operation if there are no events to upload.
+                    // 5. Stop operation if there are no events to upload.
                     return hasEvents.boolValue;
                   })
       .thenOn(self.uploaderQueue,
               ^FBLPromise<GDTCORUploadBatch *> *(id result) {
-                if (self.isCancelled) {
-                  return nil;
-                }
-
-                // 4. Fetch events to upload.
+                // 6. Fetch events to upload.
                 GDTCORStorageEventSelector *eventSelector = [self eventSelectorTarget:target
                                                                        withConditions:conditions];
                 return [storage batchWithEventSelector:eventSelector
                                        batchExpiration:[NSDate dateWithTimeIntervalSinceNow:600]];
               })
+      .thenOn(self.uploaderQueue,
+              ^FBLPromise<GDTCORUploadBatch *> *(GDTCORUploadBatch *batch) {
+                // 7. Add metrics to batch if metrics collection is supported.
+                if (![self.metricsController isMetricsCollectionSupportedForTarget:target]) {
+                  return [FBLPromise resolvedWith:batch];
+                }
+
+                return [self batchByAddingMetricsEventToBatch:batch forTarget:target];
+              })
       .validateOn(self.uploaderQueue,
-                  ^BOOL(GDTCORUploadBatch *batch) {
-                    // 5. Validate batch.
-                    return batch.batchID != nil && batch.events.count > 0;
+                  ^BOOL(id result) {
```

Comment:
Any way to get better type safety here than `id`?

Reply:
Done in [5324cea](https://github.com/google/GoogleDataTransport/pull/52/commits/5324cea052b93da71cc87b497fff065a840c7543). Discussed offline.